### DISCUSSION TITLE
Replace precompiled "script" code

### DISCUSSION THREAD

**Discussion by GravisZro** [DescentDevelopers/Descent3#206] [Ideas]
The existing method for adding specialized behavior is in the `scripts` directory which are compiled into DLLs, included in the `.hog` file, and then loaded upon request.

### Problems
* Architecture specific
* Zero security (no restrictions whatsoever)

### Solution
Replace `scripts` code with actual scripting. It has a very limited scope, so it shouldn't be too difficult.

### Implementation
There are [a LOT of possible library options](http://dbohdan.github.io/embedded-scripting-languages/) to choose from but I think it would be easiest to use [PicoC](https://gitlab.com/zsaleeba/picoc) (a platform agnostic C interpreter) as it should require the least amount of changes to `scripts` code. My second choice would be using libtcc from [TCC](https://bellard.org/tcc/) but it only supports a handful of platforms because it's an actual compiler.

**Comment by Jayman2000** [comment]
Previously, @kevinbentley suggested that we use WebAssembly for this purpose: <https://github.com/DescentDevelopers/Descent3/issues/9#issuecomment-2059801030>. I like the idea of using WebAssembly because it allows us to use whatever language we want (as long as it can be compiled to WebAssembly).

**Reply by GravisZro** [reply]
That sounds like a great idea as bytecode will execute much faster.

I did a bit of digging on github and narrowed my own choice down to two options:
* [Fizzy](https://github.com/wasmx/fizzy)
* [Wasm3](https://github.com/wasm3/wasm3)

I think Fizzy is the winner though because it seems to be more focused on correctness and is actively maintained. Unfortunately, the Wasm3's author's house was destroyed by Russians and isn't actively maintaining the project. I specifically opted for an interpreter because JIT engines aren't portable and they are too easy to break out of.  Also, the amount of code being executed is minimal so execution speed is not a huge concern.

Does anyone else know of any exceptional wasm bytecode engines we could use?
